Ai strategies implementation

baatochan opened the pull request:
Closes #16

kburzala committed:
Add alwaysHit and alwaysLeave AI options

baatochan committed:
Add backbone for ai methods

baatochan committed:
Implement solo strategy

kburzala committed:
Missing chosenPawn in checkIfHittingIsPossible()

baatochan committed:
Implement balanced ai strategy

baatochan committed:
Differentiate distance between being on start position vs being in home

baatochan committed:
Change if statement into helper method

baatochan committed:
Implement random ai strategy

kburzala committed:
Fix alwaysLeave condition (alwaysHit is blocked on distanceFromStart)

baatochan committed:
Merge branch 'always-options' into ai-strategies-implementation

baatochan committed:
Fix formatting

baatochan committed:
Refactor random strategy to not to get deadlocked

baatochan committed:
Merge two similar methods for leaving home

baatochan committed:
Fix always hit strategy modifier

baatochan committed:
Add debug prints with switch

baatochan committed:
Unify prints

baatochan commented:
Also closes #17 #18 and #1

baatochan committed:
Add hacky workaround for incorrect order of players on the main screen

baatochan committed:
Rename playerId var to be more meaningful

kburzala commented on scripts/Player.gd:
```diff
@@ -134,3 +155,107 @@ func selectPawnUsingFallbackStrategy():
+
+func fallbackToFallbackStrategy(strategyName):
+	print("ERROR: "  + str(strategyName) + " strategy had a problem and stopped working, using fallback strategy")
+	printDebug(true)
```
Is it hardcoded always-print debug output?

kburzala replied:
Other's execetued with default false value.

baatochan replied:
Yes, because this code happens when there is an ERROR (os should it be WARNING as technically the situation should never happen but IS recoverable?). Anyway the debug info is printed here always because it is additional info to the ERR/WRN print.

kburzala commented on scripts/Player.gd:
```diff
@@ -134,3 +155,107 @@ func selectPawnUsingFallbackStrategy():
+func selectTheNearestPawn():
+	var status = getPawnStatus()
+	var nearestDistance = 40
+	var nearestPawn = -1
```
Is '-1' safe? What if two pawns leave home and the second of them be placed on -2 field.

Probably better to put something like 'undefined' string value. I hope there won't be any value conflicts.

baatochan replied:
Nearest pawn is the ID of the nearest pawn, not its position so the -1 is safe as the valid pawn IDs are 1-4. It could be null or undef as well tho. I will fix that.

baatochan replied:
Changed it to null for better readability that the value is not set

kburzala reviewed (commented):
One debug print and usage of -1 value commented.

baatochan committed:
Change default value of not selected pawn id in selector methods

Both selector methods - select nearest and furthest pawn methods had not selected pawn id set as -1, null is better option for readability

baatochan commented:
I assume that because the uni project has ended @kburzala is not interested in this project anymore so I'm gonna merge it by myself
